Heatmap decompose db retrieval

fredmanglis opened the pull request:
#### Description

The commits in this pull request are a beginning is decomposing the code retrieving data from the database (in GN1), from the computation and presentation.

Here, there are functions that are to be used in the retrieval of data from the database, and the tests that check some very basic things about them.

Granted, most of the work in decomposing the db-retrieval code is not complete, but frequent merging means the integration of the code will be in small bits which will be easier to deal with in case of conflicts.

#### How should this be tested?

At this point, I do not think there is much to test, however, for anyone more versed in the system than I am (everyone else?), you can call these functions in simple scripts to see whether you get the data you expect.

#### Any background context you want to provide?

This commit tries to decompose database access code in the following modules:

- https://github.com/genenetwork/genenetwork1/blob/master/web/webqtl/base/webqtlDataset.py
- https://github.com/genenetwork/genenetwork1/blob/master/web/webqtl/base/webqtlTrait.py

this is ongoing work.

#### What are the relevant pivotal tracker stories?

https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi

#### Screenshots (if appropriate)

N/A

#### Questions

No questions, but feel free to question me about all this should you need to.

fredmanglis committed:
Retrieve 'ProbeSet' trait name

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi

* gn3/db/traits.py: new function (retrieve_probeset_trait_name)
* tests/unit/db/test_traits.py: test(s) for new function

  Add a function to retrieve the name of a 'ProbeSet' trait in a manner
  similar to genenetwork1's retrieval of the same, as implemented here
  https://github.com/genenetwork/genenetwork1/blob/master/web/webqtl/base/webqtlDataset.py#L140-154

  Unlike in genenetwork1, we do not mutate an object, instead, we return the
  values as retrieved from the database, and the caller will deal with the
  returned values as appropriate.

fredmanglis committed:
Make name retrieval more general

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi

* gn3/db/traits.py: make function more general
* tests/unit/db/test_traits.py: parametrize the tests

  Make the name retrieval more general for the different types of traits by
  changing the column specification and table as appropriate.

fredmanglis committed:
Retrieve trait information

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi

* gn3/db/traits.py: add functions to retrieve traits information
* tests/unit/db/test_traits.py: add tests for new function

  Add functions to retrieve traits information as is done in genenetwork1
  https://github.com/genenetwork/genenetwork1/blob/master/web/webqtl/base/webqtlTrait.py#L397-L456

  At this point, the data retrieval functions are probably incomplete, as
  there is more of the `retrieveInfo` function in GN1 that has not been
  considered as of this commit.

BonfaceKilz reviewed (changes requested):
@fredmanglis This LGTM! +1 for the testcases. Before I merge, could you add type annotations to your functions? For  `conn`, atm I just use "Any"...

fredmanglis committed:
Add type annotations to the function

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi

* Add some type annotations to the functions to reduce the chances of bugs
  creeping into the code.

fredmanglis committed:
Add partial type annotations for slink module

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi

* Add some type annotations for the `nearest` function.
* Leave some comments regarding the issues experienced when trying to add some
  typing annotations to the function to help with future endeavours of the
  same.

BonfaceKilz commented:
Will fix Pylint Errors in another PR!

BonfaceKilz commented:
Thanks @fredmanglis for this!

zsloan commented:
Btw I don't think GN2 uses the Temp table for storing temp traits (though GN1 did). I think it just stores them in Redis when a user adds them to a collection.

robwwilliams commented:
Let me know if you need a tour of any of the heatmap functions. In the past
all traits rum by heatmap had to be sourced from one and only one group.
This makes sense for the hierarchical clustering which needs common
individual, but it does not make sense for purely mapping functions
provided data is from one species.

It is now possible to put trait data for many groups snd even several
species into a single Collection.

If all traits SELECTED in collection are from a single group then all
heatmap functions should be possible (even a user supplied eigentrait from
the same group).

If all traits SELECTED from the same species but multiple groups then only
the mapping function of heatmap should be enabled.

If SELECTED traits include two or more species then user should be given a
error message and told that

“This function is currently only implemented for data sets all derived from
a single species. Please restrict your selected traits to one species for
heatmapping and to only one group for hierarchically-ordered heatmapping.”

On Thu, Jul 29, 2021 at 7:50 AM Muriithi Frederick Muriuki <
***@***.***> wrote:

> Description
>
> The commits in this pull request are a beginning is decomposing the code
> retrieving data from the database (in GN1), from the computation and
> presentation.
>
> Here, there are functions that are to be used in the retrieval of data
> from the database, and the tests that check some very basic things about
> them.
>
> Granted, most of the work in decomposing the db-retrieval code is not
> complete, but frequent merging means the integration of the code will be in
> small bits which will be easier to deal with in case of conflicts.
> How should this be tested?
>
> At this point, I do not think there is much to test, however, for anyone
> more versed in the system than I am (everyone else?), you can call these
> functions in simple scripts to see whether you get the data you expect.
> Any background context you want to provide?
>
> This commit tries to decompose database access code in the following
> modules:
>
>    -
>    https://github.com/genenetwork/genenetwork1/blob/master/web/webqtl/base/webqtlDataset.py
>    -
>    https://github.com/genenetwork/genenetwork1/blob/master/web/webqtl/base/webqtlTrait.py
>
> this is ongoing work.
> What are the relevant pivotal tracker stories?
>
>
> https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/clustering.gmi
> Screenshots (if appropriate)
>
> N/A
> Questions
>
> No questions, but feel free to question me about all this should you need
> to.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/genenetwork/genenetwork3/pull/31
> Commit Summary
>
>    - Retrieve 'ProbeSet' trait name
>    - Make name retrieval more general
>    - Retrieve trait information
>
> File Changes
>
>    - *M* gn3/db/traits.py
>    <https://github.com/genenetwork/genenetwork3/pull/31/files#diff-e19b79914adc9a806efdc1df524478a0b2152b8174737e01221f1b5ebc5f8770>
>    (153)
>    - *A* tests/unit/db/test_traits.py
>    <https://github.com/genenetwork/genenetwork3/pull/31/files#diff-67e1b8cb9f0c59565a92afaacf95bc4a369b9799f18b2cb2d15ca1853606fa25>
>    (114)
>
> Patch Links:
>
>    - https://github.com/genenetwork/genenetwork3/pull/31.patch
>    - https://github.com/genenetwork/genenetwork3/pull/31.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork3/pull/31>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7FH6P6WIWVQHRSKTDLT2DT2BANCNFSM5BFXNAEQ>
> .
>
-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: ***@***.***  or  ***@***.***

arunisaac commented:
Sorry, my response is late. But, I don't think we should generalize SQL
queries with extensive string interpolation like you do in
9b66f428f341bc047030126ba1e4cc405a34570c. It assumes too much about the
structure and name of the tables, and makes the code harder to read. Ideally,
we should not use string interpolation at all and move to prepared SQL
statements. See, for example,
https://en.wikipedia.org/wiki/Prepared_statement#Python_DB-API

fredmanglis commented:
@arunisaac 

> But, I don't think we should generalize SQL queries with extensive string interpolation like you do in 9b66f42

I agree: For most part, I was trying to reproduce what was in GN1 without too much repetition.

After thinking some more about this, and looking at what is required so far, I think I might change the whole thing, and fetch **ONLY** the data that is required for the clustering and heatmap work.

The original code in GN1 fetches a lot of data, because the objects are generic, and can be used in multiple places in the code-base. My new-ish approach, on the other hand, is to only fetch the data required for each "task" or "feature", and thus, it is unnecessary to try and replicate what GN1 is doing as regards fetching the data.

fredmanglis commented:
@robwwilliams 

> Let me know if you need a tour of any of the heatmap functions.

That sounds good, and might be helpful to better grasp what is needed.

> In the past all traits rum by heatmap had to be sourced from one and only one group.
> This makes sense for the hierarchical clustering which needs common
> individual, but it does not make sense for purely mapping functions
> provided data is from one species.
>
> It is now possible to put trait data for many groups snd even several
> species into a single Collection.
>
> If all traits SELECTED in collection are from a single group then all
> heatmap functions should be possible (even a user supplied eigentrait from
> the same group).
>
> If all traits SELECTED from the same species but multiple groups then only
> the mapping function of heatmap should be enabled.
>
> If SELECTED traits include two or more species then user should be given a
> error message and told that
> 
> “This function is currently only implemented for data sets all derived from
> a single species. Please restrict your selected traits to one species for
> heatmapping and to only one group for hierarchically-ordered heatmapping.”

I did not understand all that (I'm still catching up with the terminology), but from what I understood, that adds a few extra requirements to the UI.

- If traits from one "group" (not entirely sure what that means in this context), is used, then all heatmap function work
- If traits from multiple groups BUT one species: only the mapping function of heatmap is used
- If more than one species is used: no heatmap functions are used - present error/notification

I'm guessing a "collection" is the data the user searches for and selects to work on, right?

zsloan commented:
I think I can help explain what Rob was saying here (though all of this can probably wait until later).

The heatmap function is run on a set of traits, either from search results or from a user-made collection. In GN1 user-made collections could only contain traits from within the same group, but GN2 does not have this limitation.

Since heatmap is basically just doing mapping for a set of traits, the only thing that is necessary is that all those traits are from the same species (since the heatmap will be across the genome, so there just need to be the same chromosomes/chromosome lengths). They can be from different groups (within the same species) because of this.

The main complication with this that Rob doesn't mention is that different groups might have different sets of markers in their genotype files. I'm not sure how best to deal with this. It might just work to take the full set of all markers and have NAs for the ones not shared across groups, but I'm not sure.

Also, I'm not sure what the plans are with heatmap, but we aren't going to want to reproduce it the same way it works in GN1 (in terms of how the figure is drawn and how the calculations are done, since I think it uses the old qtlreaper). I think that my extremely basic implementation of it in GN2 uses this d3 figure created by Karl Broman for visualization - https://kbroman.org/d3panels/assets/test/lodheatmap/ (this example is time series, but the Y axis can also just be the different traits being mapped).

robwwilliams commented:
A bit more clarification that may help:

1. The search menu has four hierarchies:

a. Species
b. Group
c. Type
d. Dataset


For example:

Species = *Mouse*
Group = *BXD Family*
Type = *Hippocampus mRNA*
Dataset = *Hippocampus Consortium M430v2 (Jun06) PDNN*

Note that members of a group share common identifiers at some level. For
example in the BXD Family—all data for BXD1, BXD2, BXD5, ... refers to the
same "genetic" type of individual (but not the same literal individual).
Some of these families are sets of clones of animals (also known as "inbred
strains" or "recombinant inbred strains"). If we have brain weight data for
the BXDs from 1985 and liver weight data for the BXDs from 2021, then we
can compute the correlations of those weights.

Therefore: Any vector of data WITHIN a family can be correlated with any
other vector of data WITHIN the same family. This is required to compute
the hierarchical clustering.

2. Yes and Collection is any set of traits that the user has assembled into
one or more sets using the ADD function.

Check out the GN menu *Collections *pulldown. To add traits/vectors to your
own collection just do a search and then click the *Add* button. For
example if you search for the string "grin2b" using the default menu
settings. you should get 8 hits.
*Select*  them, then *Add* them to your first *Collection*.
Select them again from your Collection (perhaps named *Default*) and now
implement the *MultiMap*, aka *HeatMap*.

Do this both in GN1 and GN2. Ideally, we will replicate ALL GN1 functions
and look in GN2 and then add a mess of cool new UX features such as *Rotate
Figure*,
*Select Rows*, *Select Columns*, *Reorder Rows*, *Reorder Columns*, *Zoom
to Chr NN, *etc.

Zach, I see we use both terms MultiMap and HeatMap in GN2. The button in
*Collections* is labeled *MultiMap* but the output and the "wait" screen
uses the term "*HeatMap*" or "*Heatmap*". We need to change all to
*MultiMap*.

Most true network analysis and synthesis tools in GN start with user
defined *Collections*.

*Collections* may in principle, and in practice, contain vectors of data
(traits) from many species and many groups, but the typical pattern is that
most users only have data for a SINGLE GROUP in any one collection. Some of
these vectors of data are generated by the user, either as eigenvectors or
even as edited versions of a trait already in the database.

IMPORTANT note to Zach and Arun and Bonface:

GeneNetwork now does NOT allow users to submit edited traits into
*Collections*. This is an interface flaw we should fix ASAP.
If and when the user edits a trait in the *Trait Data and Analysis* form
they expect that when they Add this modified trait to their collection that
they are saving the modified version, not the original


Zach, this is pretty important and a bit tricky since we have to label and
save and use the user-modified version of the trait (a hash would be good).


Our GN code for almost all network analyses needs to evaluate and
categorize the structure and heterogeneity of the user's collection before
trying to invoke a function such as *HeatMap*.

Try it yourself from the user's perspective in GN2 and in GN1 to understand
the *Heatmap* function.



On Mon, Aug 2, 2021 at 11:29 PM Muriithi Frederick Muriuki <
***@***.***> wrote:

> @robwwilliams <https://github.com/robwwilliams>
>
> Let me know if you need a tour of any of the heatmap functions.
>
> That sounds good, and might be helpful to better grasp what is needed.
>
> In the past all traits rum by heatmap had to be sourced from one and only
> one group.
> This makes sense for the hierarchical clustering which needs common
> individual, but it does not make sense for purely mapping functions
> provided data is from one species.
>
> It is now possible to put trait data for many groups snd even several
> species into a single Collection.
>
> If all traits SELECTED in collection are from a single group then all
> heatmap functions should be possible (even a user supplied eigentrait from
> the same group).
>
> If all traits SELECTED from the same species but multiple groups then only
> the mapping function of heatmap should be enabled.
>
> If SELECTED traits include two or more species then user should be given a
> error message and told that
>
> “This function is currently only implemented for data sets all derived from
> a single species. Please restrict your selected traits to one species for
> heatmapping and to only one group for hierarchically-ordered heatmapping.”
>
> I did not understand all that (I'm still catching up with the
> terminology), but from what I understood, that adds a few extra
> requirements to the UI.
>
>    - If traits from one "group" (not entirely sure what that means in
>    this context), is used, then all heatmap function work
>    - If traits from multiple groups BUT one species: only the mapping
>    function of heatmap is used
>    - If more than one species is used: no heatmap functions are used -
>    present error/notification
>
> I'm guessing a "collection" is the data the user searches for and selects
> to work on, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork3/pull/31#issuecomment-891514564>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7GCODXSI6JZ5KV7PCLT25WENANCNFSM5BFXNAEQ>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: ***@***.***  or  ***@***.***

robwwilliams commented:
Zach, great points.

When the user has data for multiple *Groups* but a single *Species* in
their *Collection* then *MultiMap* should use a centimorgan scale that does
not care about individual markers per se. That will be easy when using
R/qtl1 and R/qtl2.

Zach's draft MultiMap is a good start, but we need

1. Enable flipping from horizontal mapping to vertical mapping
2. Enable users to modify the Multimap output by changing the order of data
and to delete or add new traits from the collection WITHOUT recomputing
from scratch.
3. Enable hierarchical clustering as done now in GN1.
4. Enable selection and zooming of part of the graph
5. Any of your additional ideas.


MultiMap could transform GN2. This is a true NETWORK tool if we do it
right. At present GN is 90% used for single trait analysis—not for true
network analysis. But MultiMap gets us to real genetic network
architecture.

When we database *GEMMA* precomputes of mapping results, *MultiMap* will
then become a totally awesome analysis/synthesis tool. It will, for
example, be possible to search and select all traits within a group with
-logP scores of >3.0 at two or more regions of the genome and then display
then in a *MultiMap*. This will open up new frontiers for genetics.




On Tue, Aug 3, 2021 at 12:26 AM zsloan ***@***.***> wrote:

> I think I can help explain what Rob was saying here (though all of this is
> can probably wait until later).
>
> The heatmap function is run on a set of traits, either from search results
> or from a user-made collection. In GN1 user-made collections could only
> contain traits from within the same group, but GN2 does not have this
> limitation.
>
> Since heatmap is basically just doing mapping for a set of traits, the
> only thing that is necessary is that all those traits are from the same
> species (since the heatmap will be across the genome, so there just need to
> be the same chromosomes/chromosome lengths). They can be from different
> groups (within the same species) because of this.
>
> The main complication with this that Rob doesn't mention is that different
> groups might have different sets of markers in their genotype files. I'm
> not sure how best to deal with this. It might just work to take the full
> set of all markers and have NAs for the ones not shared across groups, but
> I'm not sure.
>
> Also, I'm not sure what the plans are with heatmap, but we aren't going to
> want to reproduce it the same way it works in GN1 (in terms of how the
> figure is drawn and how the calculations are done, since I think it uses
> the old qtlreaper). I think that my extremely basic implementation of it in
> GN2 uses this d3 figure created by Karl Broman for visualization -
> https://kbroman.org/d3panels/assets/test/lodheatmap/ (this example is
> time series, but the Y axis can also just be the different traits being
> mapped).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork3/pull/31#issuecomment-891539359>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7GWHIFCYKD77NQZOH3T254YBANCNFSM5BFXNAEQ>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: ***@***.***  or  ***@***.***

arunisaac commented:
>> But, I don't think we should generalize SQL queries with extensive
>> string interpolation like you do in 9b66f42
>
> I agree: For most part, I was trying to reproduce what was in GN1
> without too much repetition.

Yeah, this is ok for now to reach your immediate goals. But, at some
point soon, it will be good to refactor the database queries. If we
don't refactor, we might find that GN3 has become just as complex as
GN1. :-P

Cheers! :-)

BonfaceKilz commented:
Muriithi Frederick Muriuki
***@***.***> anaandika:

> @arunisaac 
>
>> But, I don't think we should generalize SQL
> queries with extensive string interpolation like
> you do in 9b66f42
>
> I agree: For most part, I was trying to reproduce
> what was in GN1 without too much repetition.
>
> After thinking some more about this, and looking
> at what is required so far, I think I might change
> the whole thing, and fetch **ONLY** the data that
> is required for the clustering and heatmap work.
>
> The original code in GN1 fetches a lot of data,
> because the objects are generic, and can be used
> in multiple places in the code-base. My new-ish
> approach, on the other hand, is to only fetch the
> data required for each "task" or "feature", and
> thus, it is unnecessary to try and replicate what
> GN1 is doing as regards fetching the data.

Function Composition! ;)

-- 
Bonface M. K. D4F09EB110177E03C28E2FE1F5BBAE1E0392253F
Free Software Activist
Humble GNU Emacs User | Bearer of scheme-y parens
Curator: <https://upbookclub.com> | Twitter: @BonfaceKilz
